core/services/headtracker/head_listener.go

```diff
@@ -0,0 +1,258 @@
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/services/eth"
+	"github.com/smartcontractkit/chainlink/core/store/models"
+	"github.com/smartcontractkit/chainlink/core/utils"
+)
+
+var (
+	promNumHeadsReceived = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "head_listener_heads_received",
+		Help: "The total number of heads seen",
+	})
+	promEthConnectionErrors = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "head_listener_eth_connection_errors",
+		Help: "The total number of eth node connection errors",
+	})
+)
+
+type Config interface {
+	EthHeadTrackerMaxBufferSize() uint
+	EthereumURL() string
+	ChainID() *big.Int
+}
+
+type HeadListener struct {
+	config           Config
+	ethClient        eth.Client
+	headers          chan *models.Head
+	headSubscription ethereum.Subscription
+	connectedMutex   sync.RWMutex
+	connected        bool
+	sleeper          utils.Sleeper
+
+	log      *logger.Logger
+	muLogger sync.RWMutex
+
+	chStop chan struct{}
+	wgDone *sync.WaitGroup
+}
+
+func NewHeadListener(l *logger.Logger,
+	ethClient eth.Client,
+	config Config,
+	chStop chan struct{},
+	wgDone *sync.WaitGroup,
+	sleepers ...utils.Sleeper,
+) *HeadListener {
+	var sleeper utils.Sleeper
+	if len(sleepers) > 0 {
+		sleeper = sleepers[0]
+	} else {
+		sleeper = utils.NewBackoffSleeper()
+	}
+	return &HeadListener{
+		config:    config,
+		ethClient: ethClient,
+		sleeper:   sleeper,
+		log:       l,
+		chStop:    chStop,
+		wgDone:    wgDone,
+	}
+}
+
+// SetLogger sets and reconfigures the log for the head tracker service
+func (hl *HeadListener) SetLogger(logger *logger.Logger) {
+	hl.muLogger.Lock()
+	defer hl.muLogger.Unlock()
+	hl.log = logger
+}
+
+func (hl *HeadListener) logger() *logger.Logger {
+	hl.muLogger.RLock()
+	defer hl.muLogger.RUnlock()
+	return hl.log
+}
+
+func (hl *HeadListener) ListenForNewHeads(handleNewHead func(ctx context.Context, header models.Head) error, connected func()) {
+	defer hl.wgDone.Done()
+	defer func() {
+		if err := hl.unsubscribeFromHead(); err != nil {
+			hl.logger().Warn(errors.Wrap(err, "HeadListener failed when unsubscribe from head"))
+		}
+	}()
+
+	ctx, cancel := utils.ContextFromChan(hl.chStop)
+	defer cancel()
+
+	for {
+		if !hl.subscribe(connected) {
+			break
+		}
+		err := hl.receiveHeaders(ctx, handleNewHead)
+		if ctx.Err() != nil {
+			break
+		} else if err != nil {
+			hl.logger().Errorw(fmt.Sprintf("Error in new head subscription, unsubscribed: %s", err.Error()), "err", err)
+			hl.headers = nil
+			continue
+		} else {
+			break
+		}
+	}
+}
+
+// This should be safe to run concurrently across multiple nodes connected to the same database
+// Note: returning nil from receiveHeaders will cause listenForNewHeads to exit completely
```

Comment:
This comment doesn't seem accurate?

Reply:
the part about multiple nodes? hmm is it no longer the case? i thought it's still concurrency-safe?

Reply:
I mean, returning nil from receiveHeaders.

It doesn't cause anything to exit, does it?

Reply:
yeah it does - it uses this second 'break' here to exit the loop, so exits listenForNewHeads which in turn unsubscribes from head and ends the goroutine.

```

	for {
		if !ht.subscribe() {
			break
		}
		err := ht.receiveHeaders(ctx)
		if ctx.Err() != nil {
			break
		} else if err != nil {
			ht.logger().Errorw("Error in new head subscription, unsubscribed", "err", err)
			ht.headers = nil
			continue
		} else {
			break
		}
	}
```
I added this comment after fixing a bug around this a month ago or so